lms/djangoapps/grades/models.py

```diff
@@ -201,44 +203,55 @@ def bulk_create(cls, course_key, block_record_lists):
             )
             for brl in block_record_lists
         ])
-        cls._update_cache(course_key, created)
+        cls._update_cache(user_id, course_key, created)
         return created
 
     @classmethod
-    def bulk_get_or_create(cls, block_record_lists, course_key):
+    def bulk_get_or_create(cls, user_id, course_key, block_record_lists):
         """
         Bulk creates VisibleBlocks for the given iterator of
-        BlockRecordList objects for the given course_key, but
+        BlockRecordList objects for the given user and course_key, but
         only for those that aren't already created.
         """
-        existent_records = cls.bulk_read(course_key)
-        non_existent_brls = {brl for brl in block_record_lists if brl.hash_value not in existent_records}
-        cls.bulk_create(course_key, non_existent_brls)
+        cached_records = cls.bulk_read(user_id, course_key)
+        # Eliminate hash values already cached for this user/course
+        # from the list of hash values to look up in a database query.
+        # The database query is necessary because there may be visible blocks
+        # seen and created when grades were computed for another user that have
+        # not been seen by this user.
+        possibly_not_created = {brl.hash_value for brl in block_record_lists if brl.hash_value not in cached_records}
+        already_created = {blocks.hashed for blocks in cls.objects.filter(hashed__in=possibly_not_created)}
```

Comment:
Nit: is there a better name than `already_created`?  Cause theoretically, the value `already_created` would also include the ones that were previously created and returned in `cached_records`, right?
Maybe: `just_created` or `recently_created` ?